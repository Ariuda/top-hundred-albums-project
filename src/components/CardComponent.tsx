import React from 'react';
import { connect } from 'react-redux';

import { updateFavorite } from '../store/action-creators/albumsActions';
import { Album } from '../entities';
import ButtonComponent from './ButtonComponent';
import { limitCharacters } from '../helpers/helpers';
import { Link } from 'react-router-dom';

interface Props {
    updateFavorite: (id: string) => void;
}

type CardComponentProps = Album & Props;

export const CardComponent: React.FC<CardComponentProps> = ({ image, title, artist, category, index, id, link, favorite, updateFavorite }) => {
    const formattedTitle = limitCharacters(title);
    const formattedArtistName = limitCharacters(artist.name);

    const renderArtistName = () => {
        if (artist.link) {
            return <a href={artist.link} target="_blank" rel="noreferrer" className="artist-name">{formattedArtistName}</a>
        }
        return <p className="artist-name">{formattedArtistName}</p>;
    };

    const renderFavoriteIcon = () => {
        if(favorite) {
            return <i className="small material-icons">favorite</i>;
        }
        return <i className="small material-icons">favorite_border</i>;
    };

    return (
        <li className="card-container col-xl-4 col-lg-6 col-md-12">
            <div className="card">
                <div className="card-top-row row">
                    <div className="col-6">
                        <a href={category.link} target="_blank" rel="noreferrer">{category.name}</a>
                    </div>
                    <div className="col-6">
                        Top #{index}
                        <div className="favorite-icon" onClick={(e) => updateFavorite(id)}>
                            {renderFavoriteIcon()}
                        </div>
                    </div>
                </div>
                <div className="card-second-row">
                    <img src={image} alt={title} />
                    <div className="card-text">
                        <p className="album-title">{formattedTitle}</p>
                        {renderArtistName()}
                    </div>
                </div>
                <div className="card-thrid-row">
                    <Link to={`/album/${id}`} className="dark-button">See more</Link>
                    <ButtonComponent text="Go to Apple Music" link={link} />
                </div>
            </div>
        </li>
    )
};

export default connect(null, { updateFavorite })(CardComponent);