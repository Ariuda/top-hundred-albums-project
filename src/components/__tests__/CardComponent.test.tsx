import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { CardComponent } from '../CardComponent';

const props = {
    image: 'test-image', 
    title: 'test-title', 
    artist: {
        name: 'test-artist',
        link: 'test-artist-link'
    }, 
    category: {
        name: 'test-category',
        link: 'test-category-link'
    }, 
    index: 0, 
    id: '99999', 
    link: 'test-link', 
    favorite: false, 
    updateFavorite: () => console.log('test'),
    price: 'test-price', 
    numberOfItems: '5', 
    releaseDate: 'test-date', 
    rights: 'test-rights'
}

it('should render the button as a link when a link is passed', () => {
    const wrapper = shallow(<CardComponent {...props} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
});