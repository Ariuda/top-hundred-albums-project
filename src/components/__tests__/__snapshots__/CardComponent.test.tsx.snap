// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the button as a link when a link is passed 1`] = `
<li
  className="card-container col-xl-4 col-lg-6 col-md-12"
>
  <div
    className="card"
  >
    <div
      className="card-top-row row"
    >
      <div
        className="col-6"
      >
        <a
          href="test-category-link"
          rel="noreferrer"
          target="_blank"
        >
          test-category
        </a>
      </div>
      <div
        className="col-6"
      >
        Top #
        0
        <div
          className="favorite-icon"
          onClick={[Function]}
        >
          <i
            className="small material-icons"
          >
            favorite_border
          </i>
        </div>
      </div>
    </div>
    <div
      className="card-second-row"
    >
      <img
        alt="test-title"
        src="test-image"
      />
      <div
        className="card-text"
      >
        <p
          className="album-title"
        >
          test-title
        </p>
        <a
          className="artist-name"
          href="test-artist-link"
          rel="noreferrer"
          target="_blank"
        >
          test-artist
        </a>
      </div>
    </div>
    <div
      className="card-thrid-row"
    >
      <Link
        className="dark-button"
        to="/album/99999"
      >
        See more
      </Link>
      <ButtonComponent
        link="test-link"
        text="Go to Apple Music"
      />
    </div>
  </div>
</li>
`;
